SSZ-QL: custom Generic Merkle Proofs building the tree and collecting the hashes in one sweep

fernantho opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, check out our contribution guide here https://docs.prylabs.network/docs/contribute/contribution-guidelines
   You will then need to sign our Contributor License Agreement (CLA), which will show up as a comment from a bot in this pull request after you open it. We cannot review code without a signed CLA.
2. Please file an associated tracking issue if this pull request is non-trivial and requires context for our team to understand. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
5. A changelog entry is required for user facing issues.
-->

**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
This PR replaces the previous PR https://github.com/OffchainLabs/prysm/pull/16121, which built the entire Merkle tree and generated proofs only after the tree was complete. In this PR, the Merkle proof is produced by collecting hashes while the Merkle tree is being built. This approach has proven to be more efficient than the one in https://github.com/OffchainLabs/prysm/pull/16121.

- **ProofCollector**: 
    - New `ProofCollector` type in `encoding/ssz/query/proof_collector.go`: Collects sibling hashes and leaves needed for Merkle proofs during merkleization.
    - Multiproof-ready design with `requiredSiblings`/`requiredLeaves` maps for registering target gindices before merkleization.
    - Thread-safe: read-only required maps during merkleization, mutex-protected writes to `siblings`/`leaves`.
    - `AddTarget(gindex)` registers a target leaf and computes all required sibling gindices along the path to root.
    - `toProof()` converts collected data into `fastssz.Proof` structure. 
    - Parallel execution in `merkleizeVectorBody` for composite elements with worker pool pattern.
    - Optimized container hashing: Generalized `stateutil.OptimizedValidatorRoots` pattern for any SSZ container type:
       - `optimizedContainerRoots`: Parallelized field root computation + level-by-level vectorized hashing via `VectorizedSha256`.
       - `hashContainerHelper`: Worker goroutine for processing container subsets.
       - `containerFieldRoots`: Computes field roots for a single container using reflection and SszInfo metadata.

- **`Prove(gindex)` method** in `encoding/ssz/query/merkle_proof.go`: Entry point for generating SSZ Merkle proofs for a given generalized index.

- **Testing**
  - Added `merkle_proof_test.go` and `proof_collector_test.go` to test and benchmark this feature.

The main outcomes of the optimizations are here:
```
❯ go test ./encoding/ssz/query -run=^$ -bench='Benchmark(OptimizedContainerRoots|OptimizedValidatorRoots|ProofCollectorMerkleize)$' -benchmem
goos: darwin
goarch: arm64
pkg: github.com/OffchainLabs/prysm/v7/encoding/ssz/query
cpu: Apple M2 Pro
BenchmarkOptimizedValidatorRoots-10         3237            361029 ns/op          956858 B/op       6024 allocs/op
BenchmarkOptimizedContainerRoots-10         1138            969002 ns/op         3245223 B/op      11024 allocs/op
BenchmarkProofCollectorMerkleize-10          522           2262066 ns/op         3216000 B/op      19000 allocs/op
PASS
ok      github.com/OffchainLabs/prysm/v7/encoding/ssz/query     4.619s
```
Knowing that `OptimizedValidatorRoots` implements very effective optimizations, `OptimizedContainerRoots` mimics them.
In the benchmark we can see that `OptimizedValidatorRoots` remain as the most performant and tit the baseline here:
- `ProofCollectorMerkleize` is **~6.3× slower**, uses **~3.4× more memory** (B/op), and performs **~3.2× more allocations**. 
- `OptimizedContainerRoots` sits in between: it’s **~2.7× slower** than `OptimizedValidatorRoots` (and **~3.4× higher B/op**, **~1.8× more allocations**), but it is a clear win over `ProofCollectorMerkleize` for lists/vectors: **~2.3× faster** with **~1.7× fewer allocations** (and essentially the same memory footprint).

The main drawback is that `OptimizedContainerRoots` can only be applied to vector/list subtrees where we don’t need to collect any sibling/leaf data (i.e., no proof targets within that subtree); integrating it into the recursive merkleize(...) flow when targets are outside the subtree is expected to land in a follow-up PR.

**Which issues(s) does this PR fix?**
Partially https://github.com/OffchainLabs/prysm/issues/15598

**Other notes for review**
In this [write-up](https://hackmd.io/@fernantho/BJbZ1xmmbg), I depict the process to come up with this solution.

Future improvements:
- Defensive check that the gindex is not too big, depicted [here]( https://github.com/OffchainLabs/prysm/pull/16177#discussion_r2671684100).
- Integrate optimizedContainerRoots into the recursive merkleize(...) flow when proof targets are not within the subtree (skip full traversal for container lists).
- Add multiproofs.
- Connect `proofCollector` to SSZ-QL endpoints (direct integration of `proofCollector` for BeaconBlock endpoint and "hybrid" approach for BeaconState endpoint).

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description with sufficient context for reviewers to understand this PR.
- [x] I have tested that my changes work as expected and I added a testing plan to the PR description (if applicable).

fernantho committed:
integrate ssz info with fastssz wrapper to recursively build the merkle tree paving the way to produce proofs

fernantho committed:
added changelog

fernantho committed:
run gazelle

fernantho committed:
Merge branch 'develop' into feat/merkle-proofs-with-fastssz-wrapper

fernantho committed:
added Jun's feedback

fernantho committed:
merged MerkleTree() and toMerkleTree() legacy

fernantho committed:
connected query merkle proofs to endpoint

fernantho committed:
wip

fernantho committed:
refactored code adding ProofCollector data type. Added benchmark against hoodi beacon state

fernantho committed:
refactored ProofCollector, added siblingsGindex method. run gazelle

fernantho committed:
removed unnecessary hash tree root return from Prove function.

fernantho committed:
fixed tests - TestHashTreeRoot no longer checks hash tree root computed for merkle proofs

fernantho committed:
refactored - splitted proof collector helper type from sszInfo "Prove()"

fernantho committed:
refactored

- renamed siblingsGindex to RegisterRequiredSiblings
- removed sha256Two in favour of gohashtree.Hash()

fernantho committed:
added/modified documentation. Improve performance & concurrency safety by added read-only fields requiredIndices and requiredLeaves + mutex for collecting hashes

fernantho committed:
added includeProof field to test struct and merged BeaconState and BeaconState_withProof tests, idem for BeaconBlock tests

fernantho committed:
added test for a list length field

fernantho committed:
replaced hoodi beacon state test for a `DeterministicGenesisState`

fernantho committed:
run gazelle

fernantho committed:
added changelog

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint

syjn99 commented on changelog/fernantho_merkle-proofs-with-fastssz-wrapper.md:
I think this file can be deleted.

syjn99 commented on beacon-chain/rpc/prysm/beacon/ssz_query.go:
```diff
@@ -95,15 +95,31 @@ func (s *Server) QueryBeaconState(w http.ResponseWriter, r *http.Request) {
+			httputil.HandleError(w, "Could not marshal response to SSZ: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
 	}
```
Here we might save several lines of code like:

```go
	var response ssz.Marshaler
	// If proof is requested, generate it and return SSZQueryResponseWithProof.
	if req.IncludeProof {
		response, err = generateSSZQueryResponseWithProof(info, path, sszObject, stateRoot, encodedState, offset, length)
		if err != nil {
			httputil.HandleError(w, "Could not compute merkle proofs: "+err.Error(), http.StatusInternalServerError)
			return
		}
	} else {
		response = &sszquerypb.SSZQueryResponse{
			Root:   stateRoot,
			Result: encodedState[offset : offset+length],
		}
	}

	responseSsz, err := response.MarshalSSZ()
	if err != nil {
		httputil.HandleError(w, "Could not marshal response to SSZ: "+err.Error(), http.StatusInternalServerError)
		return
	}
```

Same for `QueryBeaconBlock` as well.

syjn99 commented on beacon-chain/rpc/prysm/beacon/ssz_query.go:
```diff
@@ -186,17 +202,72 @@ func (s *Server) QueryBeaconBlock(w http.ResponseWriter, r *http.Request) {
+
+// generateSSZQueryResponseWithProof is a generic helper that generates a merkle proof for an SSZ query
+// and returns the response with proof included.
+func generateSSZQueryResponseWithProof(
```
I think this function receives too many arguments (e.g., `encodedObject`, `sszObject`, etc.) for abstraction. If we really need a common function, it can only return `*sszquerypb.SSZQueryProof` and construct the response struct with the proof pointer.

fernantho replied:
Also agreed. It's cleaner now. There was no need for so many arguments in the function

syjn99 commented on beacon-chain/rpc/prysm/beacon/ssz_query_test.go:
```diff
@@ -25,39 +26,126 @@ import (
 	require.NoError(t, err)
-	require.NoError(t, st.UpdateBalancesAtIndex(0, 42000000000))
 
 	tests := []struct {
```
Seems like we need `name` entry for each test, as I intentionally used `path` as a name for each unit test. For example, `.slot with/without proof`. Could you do this?

fernantho replied:
Yes, combined the `path` with `includeProof` field:
```
=== RUN   TestQueryBeaconState/.slot_without_proof
=== RUN   TestQueryBeaconState/.slot_with_proof
```

```
=== RUN   TestQueryBeaconBlock/.slot_without_proof
=== RUN   TestQueryBeaconBlock/.slot_with_proof
```

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,700 @@
+
+// toProof converts the collected siblings and leaves into a fastssz.Proof structure.
+// Current behavior expects a single target leaf (single proof).
+func (pc *ProofCollector) toProof() *fastssz.Proof {
```
```suggestion
func (pc *ProofCollector) toProof() (*fastssz.Proof, error) {
```

Current `toProof` method might hide the error like "Missing sibling", so I'd say we can let the caller decide when unwanted `*fastssz.Proof`is returned.

fernantho replied:
Correct! I better return an error in this `toProof` before if the Sibling was not collected rather than hiding it up until the verifier attempts to verify the proofs

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,700 @@
+// collectSibling stores the hash for a sibling node identified by gindex.
+// It only stores the hash if gindex was pre-registered via AddTarget (present in requiredSiblings).
+// Writes to the collected siblings map are protected by the collector mutex.
+func (pc *ProofCollector) collectSibling(gindex uint64, hash [32]byte) {
```
Sorry for nitpicking: this method can be placed under `collectLeaf` for readability

fernantho committed:
combined tests' path and includeProof fields for a more explicit name

fernantho committed:
removed outdated changelog

fernantho committed:
toProof does not hide missing sibling error anymore. Reordered putLittleEndian function

fernantho committed:
refactored generateSSZQueryResponseWithProof into generateSSZQueryProof and reduced input arguments

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,699 @@
+
+	proof := &fastssz.Proof{}
+	if len(pc.leaves) == 0 {
+		return proof, nil
```
Is returning a empty proof considered as normal? Or shouldn't we create an error here as this isn't expected?

fernantho replied:
It is better tor eturn an error as it is not the normal behaviour and it would encapsulate the error until the use tries to verify the proof.

fernantho committed:
return error when calling toProof() with 0 leaves

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,699 @@
+		chunks = make([][32]byte, length)
+
+		// Parallel execution
+		workerCount := runtime.GOMAXPROCS(0) * 2
```
Is there any reason for multiplying two here?

fernantho replied:
Removing this. In this case, it is more efficient to not multiply by two. Test times got reduced by 0.1-0.2seconds:
```
--- PASS: TestQueryBeaconState (10.52s)
    --- PASS: TestQueryBeaconState/.slot_without_proof (0.14s)
    --- PASS: TestQueryBeaconState/.latest_block_header_without_proof (0.11s)
    --- PASS: TestQueryBeaconState/.validators_without_proof (0.23s)
    --- PASS: TestQueryBeaconState/.validators[0]_without_proof (0.13s)
    --- PASS: TestQueryBeaconState/.validators[0].withdrawal_credentials_without_proof (0.16s)
    --- PASS: TestQueryBeaconState/.validators[0].effective_balance_without_proof (0.11s)
    --- PASS: TestQueryBeaconState/.slot_with_proof (1.13s)
    --- PASS: TestQueryBeaconState/.latest_block_header_with_proof (1.15s)
    --- PASS: TestQueryBeaconState/.validators_with_proof (1.13s)
    --- PASS: TestQueryBeaconState/len(validators)_with_proof (1.13s)
    --- PASS: TestQueryBeaconState/.validators[0]_with_proof (1.13s)
    --- PASS: TestQueryBeaconState/.validators[0].withdrawal_credentials_with_proof (1.34s)
    --- PASS: TestQueryBeaconState/.validators[0].effective_balance_with_proof (1.13s)
    ```

fernantho replied:
Although increasing the `jobs` factor (`jobs := make(chan int, workerCount*32)`) from 16 to 32 improves the performance, it is kept as 16 as a conservative first approach
- `jobs := make(chan int, workerCount*32)` logs:
  ```bash
  --- PASS: TestQueryBeaconState/.slot_with_proof (1.06s)
      --- PASS: TestQueryBeaconState/.latest_block_header_with_proof (1.10s)
      --- PASS: TestQueryBeaconState/.validators_with_proof (1.19s)
      --- PASS: TestQueryBeaconState/len(validators)_with_proof (1.14s)
      --- PASS: TestQueryBeaconState/.validators[0]_with_proof (1.07s)
      --- PASS: TestQueryBeaconState/.validators[0].withdrawal_credentials_with_proof (1.12s)
      --- PASS: TestQueryBeaconState/.validators[0].effective_balance_with_proof (1.10s)
   ```

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,699 @@
+	pc.collectLeaf(lengthHashGindex, chunks[1])
+
+	if err := gohashtree.Hash(chunks, chunks); err != nil {
+		return [32]byte{}
```
We can also return error here, so that we don't have to hide `gohashtree.Hash()` call.

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,699 @@
+// Returns:
+// - [32]byte: mixed-in Merkle root (or zero value on hashing error).
+func (pc *ProofCollector) mixinLengthAndCollect(currentGindex uint64, chunks [][32]byte) [32]byte {
+	dataRootGindex := currentGindex * 2
```
```go
	dataRoot := chunks[0]
	lengthHash := chunks[1]
```

We can assign a couple of variable for clarity here

syjn99 reviewed (commented):
Ended the first round of review, I think overall this implementation is really nice and practical!

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint

fernantho committed:
introduced more feedback from Jun.

- mixinLengthAndCollect does not hide error
- workerCount sticks to number of logical CPU

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint

rkapka commented on changelog/fernantho_connect-merkle-proofs-with-endpoint.md:
```diff
@@ -0,0 +1,7 @@
+### Added
+
+- Added New ProofCollector type. It collects sibling hashes and leaves needed for Merkle proof during the merkleization.
+  - ProofCollector uses goroutines for vector merkleization. ProofCollector is thread-safe.
+  - ProofCollector returns `fastssz.Proof` enabling the usage of `fastssz` `Verify` function.
+- Integrated Merkle proofs into ssz-query API endpoints
+- Added testing covering the production of Merkle proof from Phase0 beacon state and benchmarked against real Hoodi beacon state (Fulu version)
```
```suggestion
### Added

- Added new ProofCollector type to ssz-query
- Integrated Merkle proofs into ssz-query API endpoints

### Ignored
- Added testing covering the production of Merkle proof from Phase0 beacon state and benchmarked against real Hoodi beacon state (Fulu version)
```
I don\t think explaining how the `ProofCollector` works should be put in the changelog. I see you have this information in the godoc of the `ProofCollector` struct, which is the ideal place for this.

rkapka commented on encoding/ssz/query/merkle_proof.go:
```diff
@@ -0,0 +1,33 @@
+// - gindex: the generalized index of the node to prove inclusion for.
+// Returns:
+// - fastssz.Proof: the Merkle proof containing the leaf, index, and sibling hashes.
+// - [32]byte: the Merkle root of the entire SSZ object.
```
```suggestion
```
You don't return this. I expect it's a leftover from an earlier design.

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,709 @@
+// - required* maps are read-only during merkleization.
+// - siblings/leaves writes are protected by mu.
+type ProofCollector struct {
+	mu sync.Mutex
```
You don't have to name the mutex. If you don't, you will be able to call mutex methods directly on the collector (e.g. `collector.Lock()`)

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,709 @@
+// Returns:
+// - [32]byte: Merkle root of the current subtree.
+// - error: any error encountered during traversal/merkleization.
+func (pc *ProofCollector) merkleize(info *SszInfo, v reflect.Value, currentGindex uint64) ([32]byte, error) {
```
The `[32]byte` return value can be removed because you don't use it anywhere

fernantho replied:
I cannot delete this `[32]byte` because it is used after each recursive call and it's key for building the whole tree. 
That `[32]byte` is just discarded when this method is the used as the entrypoint as it is not needed to build the proof

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,709 @@
+// collectLeaf checks if the given gindex is a required leaf for the proof,
+// and if so, stores the provided leaf hash in the collector.
+func (pc *ProofCollector) collectLeaf(gindex uint64, leaf [32]byte) {
+	if _, ok := pc.requiredLeaves[gindex]; !ok {
```
Is this safe? Reading the map when another routine could be modifying it?

fernantho replied:
I made a few iteration of this ProofCollector type because of this. 
The first approach was something like the following:
```golang
type ProofCollector struct {
	// Collected hashes
	siblings map[uint64][32]byte
	leaves   map[uint64][32]byte
}
```
It was totally unsafe.

The second approach introduced a mutex, but this was highly inefficient and proofs took 5seconds or more. All read and write operation locked those fields...
```golang
type ProofCollector struct {
	sync.Mutex

	// Collected hashes
	siblings map[uint64][32]byte
	leaves   map[uint64][32]byte
}
```
The third approach introduced also the required* fields that are read-only during merkleization. I should enforce this read-only behaviour instead just commenting it
```golang
type ProofCollector struct {
	sync.Mutex

	// Required gindices (registered before merkleization)
	requiredSiblings map[uint64]struct{}
	requiredLeaves   map[uint64]struct{}

	// Collected hashes
	siblings map[uint64][32]byte
	leaves   map[uint64][32]byte
}
```

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,709 @@
+
+// Merkleizers and proof collection methods
+
+// merkleize recursively traverses an SSZ info  and computes the Merkle root of the subtree.
```
```suggestion
// merkleize recursively traverses an SSZ info and computes the Merkle root of the subtree.
```

fernantho committed:
Update changelog/fernantho_connect-merkle-proofs-with-endpoint.md

Co-authored-by: Radosław Kapka <radoslaw.kapka@gmail.com>

fernantho committed:
Update encoding/ssz/query/merkle_proof.go

Co-authored-by: Radosław Kapka <radoslaw.kapka@gmail.com>

fernantho committed:
removed the name of the ProofCollector sync.Mutex field

fernantho committed:
Update encoding/ssz/query/proof_collector.go

Co-authored-by: Radosław Kapka <radoslaw.kapka@gmail.com>

rkapka commented on encoding/ssz/query/merkle_proof.go:
```diff
@@ -0,0 +1,32 @@
+	// info.source is guaranteed to be valid and dereferenced by AnalyzeObject
+	v := reflect.ValueOf(info.source).Elem()
+
+	if _, err := collector.merkleize(info, v, 1); err != nil {
```
Can you add a comment explaining why we use `1` here?

rkapka commented on encoding/ssz/query/merkle_proof.go:
```diff
@@ -0,0 +1,32 @@
+	}
+
+	collector := NewProofCollector()
+	collector.registerRequiredSiblings(gindex)
```
This doesn't necessarily have to be in this PR, but we might want to add a defensive check that the `gindex` is not too big (if possible, I expect it's non-trivial to calculate the maximum possible value). I wonder if something like this would panic or loop infinitely, or some other weird behavior occurred.
```
type S struct {
    x uint64
}

s := &S{}
s.Prove(10)
```

fernantho replied:
Agreed with this. We should add another defensive check for this. I’ll leave it for another PR.

One of the ideas we discussed in the past was to compute and store a generalized index for the subtree in the SSZInfo object. With that, computing the maximum possible value would be straightforward.

Also, as a short-term shortcut, we could add a defensive check when validating the request inputs at the endpoint, since we know we only issue proofs for BeaconState or BeaconBlock, and each of them has its own endpoint.

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,709 @@
+	root := pc.MerkleizeVectorAndCollect(fieldRoots, currentGindex, uint64(depth))
+
+	// If the container root itself is the target
+	pc.collectLeaf(currentGindex, root)
```
I think you can redesign this a little bit. If the root is the target, then you don't need to collect anything and can simply compute the root with `info.source.HashTreeRoot()` at the start of the function and return early. The only prerequisite is knowing whether the root is the target, but that can be done easily by inspecting `pc.requiredLeaves`. The full thing would look something like this:
```
func (pc *ProofCollector) merkleizeContainer(info *SszInfo, v reflect.Value, currentGindex uint64) ([32]byte, error) {
    // If the container root itself is the target
    if _, ok := pc.requiredLeaves[gindex] {
        root, err := info.source.HashTreeRoot()
        if err != nil {
            return []32byte{}, err
        }
        pc.collectLeaf(currentGindex, root)
        return root, nil
    }

     // the rest of the function
    (...)
}
```
One assumption here is that all containers will be SSZ containers, but this is true for ssz-query.

fernantho replied:
It's a great optimization because we bypass the generic merkleization algorithm entirely and delegate to the sszgen HashTreeRoot(), which is far more efficient. Implemented!

rkapka replied:
We can remove this line now, right?

fernantho replied:
Resolving this comment as it was already removed

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,709 @@
+
+// packBasicElementsToChunks packs basic type elements into 32-byte chunks.
+// Returns slice of chunks as [32]byte arrays.
+func packBasicElementsToChunks(elemInfo *SszInfo, v reflect.Value, length int) [][32]byte {
```
Is it possible to reuse `ssz.PackByChunk` instead of creating this function? You could maybe move all the ssz-query specific logic to `merkleizeVectorBody` so that the existing function can be used?

fernantho replied:
Yes, I also implemented this suggestion. I moved all the serializing logic to `merkleizeVectorBody` to call `ssz.PackByChunk`.

fernantho committed:
clarified root gindex is 1

fernantho committed:
refactored merkleizeContainer to bypass generic merkleization in favor of sszgen HashTreeRoot when target leaf is the currentGindex

fernantho committed:
removed packBasicElementsToChunks in favor of serializing in merkleizeVectorBody and using ssz.PackByChunk

fernantho committed:
run gazelle

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,704 @@
+
+// putLittleEndian writes an unsigned integer value in little-endian format.
+// Supports sizes 1, 2, 4, or 8 bytes for uint8/16/32/64 respectively.
+func putLittleEndian(dst []byte, val uint64, size int) {
```
Can you move this to `encoding/bytesutil/integers.go`?

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,704 @@
+			buf := make([]byte, elemSize)
+			elem := v.Index(i)
+			if elemInfo.sszType == Boolean {
+				if elem.Bool() {
```
This can be squashed into `if elemInfo.sszType == Boolean && elem.Bool()`

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,704 @@
+	// For packed-basic vectors: leaves == fixed chunk count.
+	var leaves uint64
+	if elemInfo.sszType.isBasic() {
+		elemLen := itemLength(elemInfo)
+		leaves = (uint64(length)*elemLen + 31) / 32
+	} else {
+		leaves = uint64(length)
+	}
```
There is a `getChunkCount` function in this package that seems to return exactly what you need

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,704 @@
+	var leaves uint64
+	if elemInfo.sszType.isBasic() {
+		// Packed-basic list: leaves is the limit in 32-byte chunks.
+		leaves = fastssz.CalculateLimit(limit, uint64(length), itemLength(elemInfo))
+	} else {
+		// Composite list: leaves is the element limit.
+		leaves = uint64(limit)
+	}
```
Just as in `merkleizeVector`, it seems that you can just use `getChunkCount`

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,704 @@
+
+	// Fixed bitvector length -> fixed number of 32-byte chunks.
+	// Note: Bitvector[0] is illegal per SSZ spec, so Length() >= 1 is guaranteed.
+	numChunks := (bv.Length() + 255) / 256
```
same here: `getChunkCount` can be used

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,704 @@
+
+	// limit is in bits; convert to fixed number of 256-bit chunks.
+	// Note: Bitlist[0] is illegal per SSZ spec, so limit >= 1 is guaranteed.
+	limitChunks := (bi.limit + 255) / 256
```
same here: `getChunkCount` can be used

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,704 @@
+// Concurrency:
+// - required* maps are read-only during merkleization.
+// - siblings/leaves writes are protected by mu.
+type ProofCollector struct {
```
There's no need to export this type or any of its methods, unless you think it's valuable to do so?

fernantho replied:
Good point. There is no need to expose this type or its methods currently, so I'll "unexport" them. In fact, I see I did not follow an homogeneous pattern for exposing merkleize functions... `MerkleizeVectorAndCollect` was exported while the remainings were not exported...
Fixed it already

fernantho committed:
refactored putLittleEndian function

-  migrated from `encoding/ssz/query/proof_collector.go` to
`encoding/bytesutil/integers.go`.
- refactored it in favour of standard library helpers.
- optimized memory allocation

fernantho committed:
integrated more feedback

- Squashed conditions into a single one
- Replaced inline chunk calculations with getChunkCount helper

fernantho committed:
unexported proofCollector type and its methods. Run gazelle.

fernantho committed:
wip

fernantho committed:
removed unnecessary line

fernantho committed:
scoped down the PR by rolling back ssz_query.go and ssz_query_test.go. Added specific tests to SSZ-QL merkle proofs.

fernantho committed:
run gazelle

syjn99 commented on changelog/fernantho_connect-merkle-proofs-with-endpoint.md:
```diff
@@ -0,0 +1,7 @@
+### Added
+
+- Added new proofCollector type to ssz-query
+- Integrated Merkle proofs into ssz-query API endpoints
```
```suggestion
```

syjn99 commented on encoding/ssz/query/merkle_proof_test.go:
```diff
@@ -0,0 +1,179 @@
+		require.Equal(t, 32, len(h), "proof hash %d is not 32 bytes", i)
+	}
+
+	return proof, gi
```
Just wondering: Do we need to return these?

fernantho replied:
No, not needed at all. Removing them

syjn99 commented on encoding/ssz/query/merkle_proof_test.go:
```diff
@@ -0,0 +1,179 @@
+	}
+}
+
+func TestProve_BeaconState(t *testing.T) {
```
I think this test can be deleted because 1) we won't run `proofCollector` for `BeaconState` and 2) this will increase overall testing duration.

fernantho replied:
Although we won't run `proofCollector` for `BeaconState` it is a complex and meaningful container to test. I haven't checked specifi differences in test coverage, but it decreases from 77% to 75.2%, so by having it we are stressing more the `proofCollector` testing (for me, test coverage metric is not similar to correct testing though).

Unlike other testing times, the ones from the `BeaconState` are not 0.
```
--- PASS: TestProve_FixedNestedContainer (0.00s)
    --- PASS: TestProve_FixedNestedContainer/.value1 (0.00s)
    --- PASS: TestProve_FixedNestedContainer/.value2 (0.00s)
--- PASS: TestProve_VariableTestContainer (0.00s)
    --- PASS: TestProve_VariableTestContainer/.leading_field (0.00s)
    --- PASS: TestProve_VariableTestContainer/.field_list_uint64[2] (0.00s)
    --- PASS: TestProve_VariableTestContainer/len(field_list_uint64) (0.00s)
    --- PASS: TestProve_VariableTestContainer/.nested.nested_list_field[1] (0.00s)
    --- PASS: TestProve_VariableTestContainer/.variable_container_list[0].inner_1.field_list_uint64[1] (0.00s)
--- PASS: TestProve_BeaconBlock (0.00s)
    --- PASS: TestProve_BeaconBlock/.slot (0.00s)
    --- PASS: TestProve_BeaconBlock/.body.randao_reveal (0.00s)
    --- PASS: TestProve_BeaconBlock/.body.attestations[0].data.slot (0.00s)
    --- PASS: TestProve_BeaconBlock/len(body.attestations) (0.00s)
--- PASS: TestProve_BeaconState (0.27s)
    --- PASS: TestProve_BeaconState/.slot (0.06s)
    --- PASS: TestProve_BeaconState/.latest_block_header (0.06s)
    --- PASS: TestProve_BeaconState/.validators[0].effective_balance (0.06s)
    --- PASS: TestProve_BeaconState/len(validators) (0.06s)
```

Run all the tests in Prysm repo to get an idea of how much this would affect the overall testing time and seems this time is negligible. 
What do you think?

syjn99 replied:
Make sense, and <0.3 sec is not that much than I expected.

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,688 @@
+		sszObj := castToSSZObject(dereferencePointer(v))
+		if sszObj == nil {
+			return [32]byte{}, fmt.Errorf("value does not implement SSZObject interface")
+		}
```
Is this change really needed? When we populate the info for a container we have this code - doesn't it assure that the source is a non-nil pointer?
```
// Dereference first in case value is a pointer.
derefValue := dereferencePointer(value)
if derefValue.Kind() != reflect.Struct {
	return fmt.Errorf("expected struct for Container type, got %v", derefValue.Kind())
}

// Reset the pointer to the new value.
sszInfo.source = castToSSZObject(derefValue)
```

fernantho replied:
Agreed. It is not needed to do it twice... I'm removing this extra code.

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,688 @@
+		return [32]byte{}, err
+	}
+
+	length := v.Len()
```
What's the difference between this and `info.VectorInfo().Length()`? It would be cleaner to always use just one, and since the info's length has been parsed by us manually, we should probably use it instead to avoid any edge cases. It also doesn't require any computation because it has been calculated up-front.

fernantho replied:
Totally agree with you, there should not be any difference between them. 

Originally, I used `info.VectorInfo().Length()`, and I still prefer it because `sszInfo` already contains all the data the `proofCollector` needs. So, it’s better to keep things consistent and rely on `sszInfo` for everything, rather than using `Len()` via reflection (even though, in this case, reflection isn’t particularly expensive).

I'm not 100% sure right now, but I remind this change was needed for a corner case (I cannot replicate now). So, there was a case when the `hashTreeRoot` from sszgen differed from the merkleization in `proofCollector`.  While debugging, I realized it was about the empty bitlist serialization. And, as I used the `v.Len()` for bitlist, I extended it for the other types that may use it, to keep a consistent pattern.

I cannot reproduce this error now, so I will rollback to `info.VectorInfo().Length()`.
I think we have prevented this from happening again with different sanity checks at analyzing ssz objects and also here in the `proofCollector` correcting wrong serialization:
```golang
// Handle zero-initialized bitlist: create a single byte with just the termination bit
if len(bitlistBytes) == 0 {
	bitlistBytes = []byte{0x01}
}
```

fernantho replied:
I'm analyzing this in depth right now. I added the following lines to test it properly. Changing `length := v.Len()` `length := int(li.Length())` makes this `TestProve_VariableTestContainer` panic. Added the following log:

```diff
func (pc *proofCollector) merkleizeList(info *SszInfo, v reflect.Value, currentGindex uint64) ([32]byte, error) {
	li, err := info.ListInfo()
	if err != nil {
		return [32]byte{}, err
	}

	length := v.Len()
+	sszInfoLength := int(li.Length())
+	if length != sszInfoLength {
+		return [32]byte{}, fmt.Errorf("list length %d exceeds limit %d", length, sszInfoLength)
+	}
```
There is a deviation from the reflection value and the sszInfo length data and the panic is caused by an out of range access in a slice:
<img width="288" height="60" alt="image" src="https://github.com/user-attachments/assets/56b194c4-e594-417d-9f1e-4dbda01e95a7" />

So, the issue was not in the bitlists as I tried to remember but in the nested lists. When analyzing a list like `NestedListField [][]byte`, the analyzer creates:
- An outer `listInfo` for the `[][]byte` with its length (e.g., 3 elements)
- A single `elemInfo`for the inner `[]byte` type (which is also a List)

When we call `merkleizeList`, if we use `sszInfo` length, it has 34 bytes length for all the lists (the length last list analyzed), even though the first and the second elements are of length 32 and 33.

So, using `v.Len()` was a way to bypass this. I think this was a known limitation in `sszInfo` that was left for a later development.

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,688 @@
+		return [32]byte{}, err
+	}
+
+	length := v.Len()
```
Same comment as in `merkleizeVector`

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,688 @@
+	// Use go-bitfield to get length and bytes with termination bit cleared
+	bl := bitfield.Bitlist(bitlistBytes)
+	data := bl.BytesNoTrim()
+	bitLength := bl.Len() // number of bits (excluding termination bit)
```
Same comment as in `merkleizeVector`

fernantho committed:
Update changelog/fernantho_connect-merkle-proofs-with-endpoint.md

Co-authored-by: Jun Song <87601811+syjn99@users.noreply.github.com>

fernantho committed:
removed legacy unnecessary return parameters from proveAndVerify helper function

fernantho committed:
using sszInfo length instead of reflection length

fernantho committed:
rolled back unneeded dereferencing, analyze object assures `source` field is non-nil pointer

fernantho committed:
rolled back unneeded dereferencing, analyze object assures `source` field is non-nil pointer. Added `OptimizedContainerRoots` to `proofCollector`

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint

fernantho committed:
refactored optimzation methods making them to belong to the proofCollector. Added brief natspec

fernantho committed:
fixed wrong call ContainerFieldRoots instead of containerFieldRoots

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,778 @@
+	for i := (n - 1) * groupSize; i < v.Len(); i++ {
+		fRoots, err := pc.containerFieldRoots(ci, v.Index(i))
+		if err != nil {
+			return [][32]byte{}, errors.Wrap(err, "could not compute validators merkleization")
```
```suggestion
			return [][32]byte{}, errors.Wrap(err, "could not merkleize container")
```

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,778 @@
+		return [][32]byte{}, nil
+	}
+
+	wg := sync.WaitGroup{}
```
I know that the `sync.WaitGroup` is what the original code uses, but I think it's worth changing it to `errgroup.WithContext` from the `"golang.org/x/sync/errgroup"` package. That will cancel processing remaining groups when one of them fails.

fernantho replied:
Switched it! It's cleaner and if any of them fails, it now early exits.

fernantho committed:
reordered collectSibling method in proofCollector and added unit testing

fernantho committed:
integrated optimizedContainerRoots in merkleizeVectorBody. Added helper functions to assess if the proofCollector hasTargetsInSubtree so optimizedContainerRoots can be used. Added unit testing

fernantho committed:
Update encoding/ssz/query/proof_collector.go

Co-authored-by: Radosław Kapka <radoslaw.kapka@gmail.com>

fernantho committed:
replaced sync.WaitGroup for errgroup.WithContext

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint

syjn99 commented on encoding/ssz/query/merkle_proof_test.go:
```diff
@@ -0,0 +1,178 @@
+	ssz "github.com/prysmaticlabs/fastssz"
+)
+
+func TestProve_FixedNestedContainer(t *testing.T) {
```
```suggestion
func TestProve_FixedTestContainer(t *testing.T) {
```

We can use `createFixedTestContainer` helper function, just like `TestProve_VariableTestContainer`. I think `FixedTestContainer` is much better than `FixedNestedContainer` for testing purpose.

fernantho replied:
Agreed, this keeps things more consistent.

syjn99 commented on encoding/ssz/query/merkle_proof_test.go:
```diff
@@ -0,0 +1,178 @@
+	require.NoError(t, err)
+	protoRoot, err := sszObj.HashTreeRoot()
+	require.NoError(t, err)
+	require.Equal(t, stateRoot, protoRoot)
```
Do we need this in this test?

fernantho replied:
This sanity is not necessary as it is already tested in `./encoding/ssz/query/query_test.go`.
I removed it.

syjn99 commented on encoding/ssz/query/proof_collector_test.go:
```diff
@@ -0,0 +1,726 @@
+	"github.com/OffchainLabs/prysm/v7/testing/require"
+)
+
+// Unit testing for ProofCollector
```
```suggestion
```

syjn99 commented on encoding/ssz/query/proof_collector_test.go:
```diff
@@ -0,0 +1,726 @@
+	}
+}
+
+func makeFixedTestContainer(seed byte) *sszquerypb.FixedTestContainer {
```
We can save quite a lot of lines using existing helper function like `createFixedTestContainer`. If we really need `seed`, you might modify the `create*` functions as well.

fernantho replied:
Thanks for the suggestion. I want to keep this as whitebox test and to use `createFixedTestContainer` I'd have to change the `package query` to `package query_test`. That would me a lot of changes and exports of methods and fields within the `ProofCollector`. 

The option I've tried to prevent these "exports" has been to move `createFixedTestContainer` and `createVariableTestContainer` to an helpers.go in testutil package `encoding/ssz/query/testutil`. The problem here is that when I import this into `proof_collector_test.go` belonging to `package query` introduces a circular dependency.

The seed is for sure not needed in these tests and, although it does not add any unmanageable complexity, I removed it.

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,844 @@
+//
+// Concurrency:
+// - required* maps are read-only during merkleization.
+// - siblings/leaves writes are protected by mu.
```
```suggestion
// - siblings/leaves writes are protected by mutex.
```

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,844 @@
+// addTarget register the target leaf and its required sibling nodes for proof construction.
+// Registration should happen before merkleization begins.
+func (pc *proofCollector) addTarget(gindex uint64) {
+	// Lock safe just in case the collector is re-used.
```
```suggestion
```

Not so sure whether this comment is necessary, feeling little bit self-explanationary comment.

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,844 @@
+
+	proof := &fastssz.Proof{}
+	if len(pc.leaves) == 0 {
+		return nil, fmt.Errorf("no leaves collected: add target leaves before merkleization")
```
```suggestion
		return nil, errors.New("no leaves collected: add target leaves before merkleization")
```

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,844 @@
+func (pc *proofCollector) registerRequiredSiblings(gindex uint64) {
+	pc.reset()
+	pc.addTarget(gindex)
+}
```
I think we can just call `addTarget` in case we need to register, which means we might not need this method at all. This seems like unnecessary wrapper as well as quite implicit that it resets the underlying `pc`.

fernantho replied:
Removed it. Yes, the wrapper is not needed. Probably by doing this, there is no need for any changes here for the multi-proofs, only in the call from `encoding/ssz/query/merkle_proof.go`.

syjn99 commented on encoding/ssz/query/merkle_proof_test.go:
```diff
@@ -0,0 +1,178 @@
+	}
+}
+
+func TestProve_BeaconBlock(t *testing.T) {
```
This test panics.

```
--- FAIL: TestProve_BeaconBlock (0.00s)
    --- FAIL: TestProve_BeaconBlock/.slot (0.00s)
panic: odd number of chunks [recovered, repanicked]
```

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,844 @@
+	if err := gohashtree.Hash(chunks, chunks); err != nil {
+		return [32]byte{}, err
+	}
+	return chunks[0], nil
```
```suggestion
	return ssz.MixInLength(dataRoot, lengthHash[:])
```

We can use explicit `MixInLength` function, and we can also change the function signature not to return error.

fernantho replied:
Agreed, `gohashtree.Hash` introduces overhead without much optimizations as it is just one hash. Added your suggestion

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,844 @@
+	pc.collectLeaf(dataRootGindex, dataRoot)
+
+	// Check if lengthHash is a leaf we need to collect
+	pc.collectLeaf(lengthHashGindex, lengthHash)
```
```suggestion
	dataRoot, lengthHash := chunks[0], chunks[1]
	dataRootGindex, lengthHashGindex := currentGindex*2, currentGindex*2+1

	pc.collectSibling(dataRootGindex, dataRoot)
	pc.collectSibling(lengthHashGindex, lengthHash)

	pc.collectLeaf(dataRootGindex, dataRoot)
	pc.collectLeaf(lengthHashGindex, lengthHash)
```

I think the function comment is well-described, so let's reduce few lines here.

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,844 @@
+
+// hashContainerHelper generalizes stateutil.hashValidatorHelper for any SSZ container type.
+// Returns a function suitable for errgroup.Go that processes a group of containers.
+func (pc *proofCollector) hashContainerHelper(ctx context.Context, ci *containerInfo, v reflect.Value, roots [][32]byte, j int, groupSize, containerFieldRoots int) func() error {
```
I think `hashContainerHelper` and `containerFieldRoots` are much closer to "helper" function, which means they don't have to be members under `ProofCollector`.

Also, this part looks little bit weird:

```go
		// Non-proof path: use a constant gindex to avoid proof bookkeeping.
		root, err := pc.merkleize(fieldInfo.sszInfo, fieldVal, 0)
```

As we're only trying to get the hash tree root of each field *without* the context of `ProofCollector`.

fernantho replied:
Yes, the call seems a bit weird. For these cases, using the `ProofCollector` seems wrong, when we do not even need to collect anything. 
This is just a shortcut to have a generic way of getting the hash tree root of all the different type of fields, even for those that do not implement the `HashTreeRoot()`, which I originally tried to use but panicked with basic types.

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,844 @@
+		if err != nil {
+			return [32]byte{}, err
+		}
+	} else {
```
I think the logic under this line is hard to read, and it seems like `errgroup` can handle lots of custom things here.

fernantho committed:
Update encoding/ssz/query/proof_collector_test.go

Co-authored-by: Jun Song <87601811+syjn99@users.noreply.github.com>

fernantho committed:
Update encoding/ssz/query/proof_collector.go

Co-authored-by: Jun Song <87601811+syjn99@users.noreply.github.com>

fernantho committed:
Update encoding/ssz/query/proof_collector.go

Co-authored-by: Jun Song <87601811+syjn99@users.noreply.github.com>

fernantho committed:
Update encoding/ssz/query/proof_collector.go

Co-authored-by: Jun Song <87601811+syjn99@users.noreply.github.com>

fernantho committed:
added Jun's feedback: replaced FixedNestedContainer in favor of FixedTestContainer and removed unnecessary sanity check

fernantho committed:
replaced gohashtree.Hash for ssz.MixInLength

rkapka commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -797,17 +800,24 @@ func (pc *proofCollector) optimizedContainerRoots(info *SszInfo, v reflect.Value
+			}
+			for k, root := range fRoots {
+				roots[(j*groupSize+i)*containerFieldRoots+k] = root
+			}
```
It would be more readable to put this logic inside the `default` case

fernantho replied:
I'll leave this one for another PR only for the optimization

fernantho committed:
extracted containerFieldRoots(...) and hashContainerHelper(...) from proofCollector.

fernantho committed:
removed optimizedContainerRoots, hashContainerHelper and containerFieldRoots. Removed the related tests

fernantho committed:
removed unnecessary seed from makeVariableTestContainer and makeFixedTestContainer functions

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint

syjn99 commented on encoding/ssz/query/proof_collector.go:
```diff
@@ -0,0 +1,708 @@
+// hasTargetsInSubtree checks if any required gindex (leaf or sibling) falls within
+// the subtree rooted at subtreeRoot (strictly inside, not the root itself).
+// A gindex g is strictly inside the subtree of r if r is a proper ancestor of g.
+func (pc *proofCollector) hasTargetsInSubtree(subtreeRoot uint64) bool {
```
It seems like this isn't now used in other parts rather than test functions, do we need to keep this?

Same for `isStrictAncestor`.

fernantho committed:
Update encoding/ssz/query/proof_collector.go

Co-authored-by: Jun Song <87601811+syjn99@users.noreply.github.com>

fernantho committed:
removed proof collector unused isStrictAncestor, hasTargetsInSubtree and removed registerRequiredSiblings in favour of just using addTarget

fernantho committed:
fixed linter issues

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint

fernantho committed:
fixed linting issues in proof_collector_test.go

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint

rkapka commented on encoding/ssz/query/proof_collector_test.go:
```diff
@@ -0,0 +1,532 @@
+	pc.addTarget(6)
+
+	elements := [][32]byte{{1}, {2}}
+	// expected := ssz.MerkleizeVector(append([][32]byte{}, elements...), 2)
```
```suggestion
```

rkapka committed:
Apply suggestion from @rkapka

fernantho committed:
Merge branch 'develop' into feat/connect-merkle-proofs-with-endpoint
